warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/facet/FacetHandler.java

```diff
@@ -0,0 +1,366 @@
+                        }
+                        
+                        Text row = createFieldValuePair(pivotTypes.getIndexedFieldValue(), facetTypes.getIndexedFieldValue(), event.getDataType());
+                        Key result = new Key(row, myCf, dateColumnQualifier, cv, event.getDate());
+                        results.put(new BulkIngestKey(facetTableName, result), sharedValue);
+                        
+                        countWritten++;
+                    }
+                    
+                    myCf = generatedCf;
+                }
+                
+                Text row = createFieldValuePair(pivotTypes.getIndexedFieldValue(), pivotTypes.getIndexedFieldValue(), event.getDataType());
+                Key result = new Key(row, reflexiveCf, dateColumnQualifier, cv, event.getDate());
+                results.put(new BulkIngestKey(facetTableName, result), sharedValue);
+            }
+        }
+        
+        for (Map.Entry<String,String> pivot : pivotMap.entries()) {
+            Key result = new Key(new Text(pivot.getKey() + NULL + pivot.getValue()), PV);
+            results.put(new BulkIngestKey(facetMetadataTableName, result), EMPTY_VALUE);
+            countWritten++;
+        }
+        contextWriter.write(results, context);
+        return countWritten;
+    }
+    
+    private Multimap<String,NormalizedContentInterface> hashEventFields(Multimap<String,NormalizedContentInterface> fields,
+                    HashTableFunction<KEYIN,KEYOUT,VALUEOUT> func) {
+        Multimap<String,NormalizedContentInterface> eventFields = HashMultimap.create();
+        for (Map.Entry<String,Collection<NormalizedContentInterface>> entry : fields.asMap().entrySet()) {
+            Collection<NormalizedContentInterface> coll = func.apply(entry.getValue());
+            if (coll != null && !coll.isEmpty()) {
+                eventFields.putAll(entry.getKey(), coll);
+            }
+        }
+        return eventFields;
+    }
+    
+    /**
+     * Create the column qualifier that includes pivotFieldValue, facetFieldValue and datatype.
+     * 
+     * @param pivotFieldValue
+     * @param facetFieldValue
+     * @param dataType
+     * @return
+     */
+    protected Text createFieldValuePair(String pivotFieldValue, String facetFieldValue, Type dataType) {
+        return new Text(pivotFieldValue + NULL + facetFieldValue + NULL + dataType.typeName());
+    }
+    
+    /**
+     * Create the column family consisting of pivotFieldName and facetFieldName
+     * 
+     * @param pivotFieldName
+     * @param facetFieldName
+     * @return
+     */
+    protected Text createColumnFamily(String pivotFieldName, String facetFieldName) {
+        return new Text(pivotFieldName + NULL + facetFieldName);
+    }
+    
+    /** TODO */
+    @Override
+    public FacetValue estimate(RawRecordContainer input) {
+        // precision value: 10, sparse set disabled.
```

Comment:
Are you avoiding using the sparse setting for HLL++ because of some of the issues with size in certain cases when serializing it to disk?

Reply:
I need to take a close look at this - the sparse setting seems a little wasteful, I'm not certain when the sparse setting becomes more efficient than the alternative (?dense?) setting.